src/Compilers/CSharp/Portable/Emitter/Model/SynthesizedPrivateImplementationDetailsStaticConstructor.cs

```diff
@@ -51,22 +52,61 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
                 body.Add(payloadInitialization);
             }
 
-            // MVID = TypeOf(PrivateImplementationDetails).Module.ModuleVersionId;
-
-            BoundStatement mvidInitialization =
-                factory.Assignment(
-                    factory.ModuleVersionId(),
-                    factory.Property(
-                        factory.Property(
-                            factory.TypeOfPrivateImplementationDetails(),
-                            WellKnownMember.System_Type__Module),
-                        WellKnownMember.System_Reflection_Module__ModuleVersionId));
-            body.Add(mvidInitialization);
+            // Initialize the module version ID (MVID) field. Dynamic instrumentation requires the MVID of the executing module, and this field makes that accessible.
+            body.Add(InitializeMVID(factory, diagnostics));
 
             BoundStatement returnStatement = factory.Return();
             body.Add(returnStatement);
 
             factory.CloseMethod(factory.Block(body.ToImmutableAndFree()));
         }
+
+        private BoundStatement InitializeMVID(SyntheticBoundNodeFactory factory, DiagnosticBag diagnostics)
+        {
+            CSharpCompilation compilation = factory.Compilation;
+            CSharpSyntaxNode syntax = factory.Syntax;
+
+            // It is necessary to reflect over a type defined in the module. The PrivateImplementationDetails class serves as well as any other type
+            // for this purpose, and happens to be available here.
+            BoundExpression typeInstance = factory.TypeOfPrivateImplementationDetails();
+
+            // Getting to the MVID requires getting to a representation of the current module. If the target platform is Desktop, the module is
+            // available as a property of a type instance. If the target platform is Portable, a type instance does not have a module property.
+            // Instead, the module is available as a property of a member info object, which is available as a base type of a type info object,
+            // which is available via a GetTypeInfo extension method.
+            //
+            // The Portable mechanism also works on Desktop, but not on versions prior to 4.5, so there is no one single mechanism that work on all platforms.
+
+            BoundExpression moduleReference;
+            Symbol system_Type__Module = compilation.GetWellKnownTypeMember(WellKnownMember.System_Type__Module);
```

Comment:
It would be good to add a comment that we are relying on SyntheticBoundNodeFactory to report any appropriate use-site diagnostics.